### PR TITLE
Robin coupling: convergence checks, automatic Robin coefficient and mass-conserving interface flux

### PR DESCRIPTION
**Pull request by philipcardiff**
Robin-Neumann FSI coupling: convergence checks, automatic Robin coefficient and interface mass conservation. This PR now includes the work first opened as #452.

## Summary

**Convergence checks and beamInCrossFlow Robin variant** (original scope of this PR)
- Robin interface residuals are added to the shared FSI convergence criteria for the fixed-relaxation, Aitken and IQN-ILS loops. Non-Robin interfaces keep the displacement-only behaviour.
- A Robin coupling variant of `beamInCrossFlow`, with docs, regression coverage and a Robin-versus-IQN-ILS verification comparison. Also cerebral aneurysm mass-flow diagnostics.

**Automatic Robin coefficient** (`elasticWallPressure`)
- New `hsModel` options. `secant` is the new default: it rescales a thickness-limited estimate, `hs = l tanh(H/l)` with a ray-cast wall thickness, using the solid and fluid interface impedances measured between FSI iterations. The fit is updated every iteration to minimise the largest predicted contraction factor (minimax).
- The previous `c_p*dt` default crashed in 4 of 9 tested configurations (thin or nearly incompressible walls, large time steps).
- The learned coefficient is written, restored on restart, and mapped. `writeDiagnostics` writes the measured impedances and the predicted and observed rates.

**Mass-conserving Robin interface flux**
- The converged flux of Tuković et al. (2018) equals the mesh (GCL) flux only in the continuum. The implementation carried each step's discrete mismatch forward through `Uf.oldTime()`, which gave leakage floors of 2–4% of the throughput. Those floors made the original absolute-leakage criterion unattainable: every 3dTube step ran to `nOuterCorr`.
- pimpleFluid and interFluid (OpenFOAM.com) now start the flux from the old wall velocity. When the fluid mesh follows the solid, they also build the flux's explicit part from the mesh flux. The converged flux then equals the mesh flux, as with a Dirichlet velocity condition. The Robin condition itself (solid inertia, added mass) is unchanged.
- A pre-existing unit error in the interFluid Robin gravity/surface-tension term is fixed.

**Convergence criteria (revised)**
- The default `robinConvergence residual` tests the displacement residual, the pressure change and, when kinematically consistent, the leakage. The leakage is normalised by the physical throughput plus the interface motion flux, with default tolerance `10*outerCorrTolerance`.
- Stalled residuals within 10× the tolerances are accepted, which avoids iterations that only reproduce inner-solver noise.
- The Robin criteria are not applied before `couplingStartTime`; previously every uncoupled HronTurek step took 100 iterations.
- `iterationError` is an optional rate-based error estimate. `fsiResiduals.dat` gains a convergence-state column.

**Recommended setup and tutorials**
- Robin-Neumann is designed for `fixedRelaxation` with `relaxationFactor 1`. A one-time warning is issued otherwise. No tested case needed or benefited from Aitken or IQN-ILS with Robin, and the mass-consistent flux requires the mesh to follow the solid.
- `fillingElasticContainer` now defaults to Robin with fixed relaxation 1 and the automatic coefficient (6.3 FSI iterations per step, against 8.7 with IQN-ILS).
- Its reference apex displacement changes from −0.479 to −0.4822. The displacement-only criterion accepted steps after about one iteration, and the converged coupling gives −0.4822 across all coupling methods and coefficient models.

## Results (OpenFOAM-v2412, mean FSI iterations per step)

| case | original #450 criteria | this PR | converged leakage |
|---|---|---|---|
| 3dTube | 100 (every step capped) | 5.9 | 8e-8 |
| beamInCrossFlow modified / original | 23.9 / 9.05 | 24.3 / 9.0 | 1e-6 / 8e-8 |
| cerebralAneurysm | 30 (every step capped) | 17.3 | 1.3e-4 |
| flexibleDamBreak (Robin variant) | 94.5 | 11.8 | 5e-9 |
| fillingElasticContainer | 21.4 | 6.3 | 1.1e-4 |

In 3dTube the Robin solution moves towards the Dirichlet-Neumann one: maximum displacement difference 0.49% → 0.12%, force difference 0.46% → 0.23%.

## Verification (OpenFOAM-v2412)

- `libsolids4FoamModels` builds.
- Regression tests pass: 3dTube, beamInCrossFlow (aitken, iqnils, robin, highOrder), cerebralAneurysm, and fillingElasticContainer with the updated reference.
- `tutorials/Alltest`: 67 cases run, none failed.
- The beamInCrossFlow Robin run in parallel (4 cores) reproduces the serial iteration counts. Restart restores the learned Robin coefficient.
- Reviewed with codex (gpt-6-astra) in four rounds, plus a Claude Code ultrareview; all findings addressed.

## Not verified / limitations

- Not compiled on OpenFOAM.org or foam-extend. Their pimpleFluid variants keep the paper's flux, so leakage is reported but not tested.
- A scalar Robin coefficient cannot converge problems with widely separated interface modes quickly. HronTurekFsi3 with Robin stays at a contraction rate of about 0.95.
- The about 2% Robin vs Dirichlet-Neumann pressure-force difference in the viscous beamInCrossFlow is not caused by leakage. It comes from the inertial wall momentum balance in the Robin flux, and is left for separate work.

@conor-moore-123 — mentioning you because of your interest in Robin-related FSI convergence checks.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01SUvzjpPgL1vhRAJpX8SBSi

### CODE REVIEW

**Comment by philipcardiff** [comment]
I also tested the new Robin convergence gates on `cerebralAneurysm` and added mass-flow diagnostics in fe2378ca.

### Test setup

- OpenFOAM v2412, serial
- initial transient through `t = 0.0003 s` (six time steps)
- gated run: current defaults (`robinPressureTolerance` and `robinFluxTolerance` inherit `outerCorrTolerance = 5e-5`)
- control run: identical case with both Robin tolerances set to `1e30`, reproducing the previous displacement-only stopping rule

### Comparison

| Quantity | Robin gates | Displacement-only | Difference |
|---|---:|---:|---:|
| Total FSI iterations (six steps) | 180 | 47 | 3.83x |
| Max displacement at `t = 0.0003 s` | 3.84523e-5 m | 3.84542e-5 m | 0.00494% |
| Normal force at `t = 0.0003 s` | -1.34797e-2 N | -1.34776e-2 N | 0.0156% |
| Absolute interface mass flow at `t = 0.0003 s` | 3.06549e-4 kg/s | 3.06857e-4 kg/s | 0.100% |

Across all six common time steps, the maximum relative differences were 0.0293% for maximum displacement and 0.191% for normal force.

Every gated time step reached `nOuterCorr = 30`. At `t = 0.0003 s`, its final pressure and flux residuals were 6.71e-5 and 3.751e-2, respectively. The displacement-only run stopped after five iterations with a flux residual of 3.752e-2. Thus the additional 25 iterations did not reduce the leakage plateau materially.

The new `surfaceFieldValue` diagnostics write signed inlet, outlet, and interface mass-flow histories, plus `interfaceAbsoluteMassFlowRate`, which prevents local interface inflow/outflow from cancelling. At `t = 0.0003 s`, the absolute interface flow is about 4.9% of the inlet mass flow.

The original ten-step regression with the gates enabled still passes its displacement and force checks, but takes 30 FSI iterations per time step. This suggests that inheriting the displacement tolerance for the Robin flux criterion is too strict for this case and should be reconsidered separately from the correctness of the diagnostic calculation.

**Comment by philipcardiff** [comment]
## Robin residuals for a flow at or near rest

Review point: can the new Robin convergence checks fail to converge when there
is no throughput and the interface does not move?

The issue is the normalising scale, not `robinFluxResidual_` itself. The
leakage residual is `gSum(mag(leak))/(maxRobinKinematicNorm_ + SMALL)`. Its
scale is the boundary throughput plus the Robin interface motion flux, and it
is reset every time step. The pressure residual has the same form, normalised
by the largest interface-pressure norm.

- **Exactly trivial solution**: the leakage and the scale are both zero, so the
  residual is 0/SMALL = 0 and the check passes. The beam case with zero inflow
  converges in 1 iteration per step.
- **Near rest**: the scale tends to zero while the leakage keeps a floor set by
  the inner solvers. In a closed or pressurised cavity the throughput is zero,
  so the scale is only the interface motion. As that decays, the relative
  residual grows even though the solution is no longer changing.
- Skipping the check when the throughput is zero would be unsafe. A closed
  system at rest with real wall leakage is exactly what the check should
  detect.

### Change (74741b01)

Two optional `fsiProperties` entries, both default 0 (off):

- `robinFluxReferenceVelocity`: the flux scale is at least `Uref * |Γ_R|`,
  where `|Γ_R|` is the Robin interface area.
- `robinPressureReference`: a pressure floor, in the units of the fluid
  solution pressure (kinematic for `pimpleFluid`, Pa for `p_rgh`). It enters
  as the L2 norm of a uniform interface pressure, so it is consistent with the
  existing pressure norm.

Both are applied as a floor on the denominator, `max(scale, reference)`, not
as a separate absolute-tolerance test. The residual histories used by the
`residual` test, the stall test and the `iterationError` estimator therefore
all see the same bounded quantity. They only take effect when the flow's own
scale is smaller, and the log then marks the residual with
`(reference flux scale)` or `(reference pressure scale)`.

No scale is imposed by default. A flow at rest has no physical velocity or
pressure scale that the code could derive, so any automatic floor would be
arbitrary. Instead, if the FSI iterations reach `nOuterCorr` with an unscaled
Robin residual above its tolerance, a one-time hint suggests the options (or
checking the inner-solver tolerances and `nOuterCorr`). The default fails
safe: it over-iterates rather than accepting real leakage.

### Checks (OpenFOAM v2412)

- Defaults unchanged: the `beamInCrossFlow` Robin regression case (t = 1 s)
  gives a byte-identical `fsiResiduals.dat` to the pre-change run.
- Near rest: `beamInCrossFlow` with zero inflow and a fixed outlet pressure of
  0.1 (the beam is loaded and then settles). The converged leakage residual
  rises from 3.6e-6 at t = 0.15 s to 1e-3 to 4e-3 by t = 4 to 5.5 s, against a
  tolerance of 5e-3. This is the expected degeneracy. With the default
  tolerance it still converges in this case, because the leakage keeps
  contracting with the iterations.
- With `robinFluxTolerance 1e-4` (restart from t = 4 s):
  - without a reference: 15 to 24 iterations per step, and the leakage limits
    convergence;
  - with `robinFluxReferenceVelocity 1e-3`: 12 to 18 iterations per step, where
    the displacement residual limits convergence and the log shows
    `(reference flux scale)`.
- Hint: with `nOuterCorr 5`, it is printed once and not repeated on later
  steps.

I could not reproduce a hard non-convergence here. That needs a leakage floor
that no longer contracts, and in this case the solid PETSc SNES reached its
absolute floor first: an inflow of 1e-8 m/s or less aborted the solid solve.
The settling case needs about 4 s of simulated time (about 10 minutes of
runtime), so I have not added it to the regression script. OpenFOAM-9 and
foam-extend are left to CI.

**Comment by philipcardiff** [comment]
**Codex review of 74741b01** (Robin reference scales): no findings. It checked the parallel reductions, the normalisation, the hint condition against the convergence logic, and API compatibility with OpenFOAM.com, OpenFOAM-9 and foam-extend-4.1.

**CI fix (0745e59a).** The foam-extend-4.1-PETSc regression job failed in `3dTube`, in the backward-restart test that #380 added to `development`. Re-reading the written `p` stopped with `keyword file is undefined` on the inlet `timeVaryingUniformFixedValue`. The cause is in solids4foam's `interpolationTable`: on foam-extend it writes `fileName` but only read `file`, so it could not read back its own output. The tutorial's `"file|fileName"` entry hides this on a fresh start. The restart test did not run in other PRs' foam-extend runs, because their main `3dTube` case stopped before the requested end time. On this branch the main case completes, so the restart test runs and fails. The dictionary constructor now accepts `fileName` when `file` is absent. Existing dictionaries that use `file` are unchanged.

**Comment by philipcardiff** [comment]
## Robin pressure-change tolerance default, and cerebralAneurysm follow-up

Two commits pushed: a library default change and the `cerebralAneurysm`
tutorial updates that motivated it.

### `robinPressureTolerance` default: `outerCorrTolerance` → `10*outerCorrTolerance`

`robinFluxTolerance` already defaulted to `10*outerCorrTolerance`, but
`robinPressureTolerance` defaulted to `outerCorrTolerance` itself. The pressure
change is normalised by the largest interface-pressure norm in the time step,
so on a case with a large mean interface pressure the *relative* residual
bottoms out at the floors of the inner solvers before it can reach a tolerance
set equal to `outerCorrTolerance`. The FSI loop then keeps iterating on a
stalled residual until the stall check releases it.

`cerebralAneurysm` showed this plainly: the pressure residual stalls at about
`6e-5` against its `5e-5` tolerance, and steps were running to 25 correctors
before the stall rule let them go. Roughly 10 correctors per time step over the
first 100 steps, against 4.5 once the Robin tolerances are loosened.

Worth noting as independent evidence: `beamInCrossFlow` had already worked
around this by hand, setting `robinPressureTolerance 1e-5` against its
`outerCorrTolerance 1e-6` — exactly the 10x factor this commit makes the
default.

**Verification.** The regression tests of every tutorial using
`elasticWallPressure` pass unchanged with the new default:

| tutorial | relies on default? | result |
|---|---|---|
| `3dTube` | yes (1e-6 → 1e-5) | PASS — max disp Δ 1.56e-09, mean force Δ 2.7e-06 |
| `fillingElasticContainer` | yes (1e-4 → 1e-3) | PASS — apex dy Δ 0.00121 |
| `beamInCrossFlow` | no, sets 1e-5 | PASS — robin and highOrder variants |
| `cerebralAneurysm` | no, sets 1e-3 | PASS — disp Δ 2.66e-07, force Δ 3.47e-04 |

### cerebralAneurysm

**Windkessel comments were reversed.** `0/fluid/p` labelled `R` as `Rp` and
`Rch` as `Rd`. Per `windkesselPressureFvPatchScalarField.H`, `R` is the distal
resistance (`Rd`) and `Rch` the proximal/characteristic one (`Rp`), and `R` is
the resistance that forms the RC time constant with the compliance. Fixed.

**`hsModel secant` replaces `constantHs 5e-4`.** secant needs no
case-specific coefficient and is also marginally cheaper here. Over the first
2000 time steps (initial transient plus systole), at otherwise identical
settings:

| `hsModel` | FSI correctors/step | solver time |
|---|---|---|
| `secant` | **2.475** | 1873 s |
| `constantHs 5e-4` | 2.622 | 1931 s |

The README documents both, since a hand-tuned constant remains a reasonable
choice when a good value is known for the geometry.

**Robin tolerances set to 1e-3**, above the new `5e-4` default. Which residual
binds turns out to be regime-dependent, which is worth knowing when reading
these numbers: the *pressure* residual binds during the initial transient,
while the *leakage* residual is what blocks the last corrector once the flow
has settled (1.25e-3 against a 1e-3 tolerance at a representative settled
step). Measured over the first 2000 steps:

| Robin tolerances | FSI correctors/step |
|---|---|
| 1e-3 (shipped) | **2.475** |
| 5e-4 (new defaults) | 2.869 |

Loosening further was measured and rejected: `robinFluxTolerance 2e-3` saves
11.5% of correctors but shifts the solution by ~1.2e-3 relative, and `5e-3`
gives no further gain over `2e-3`. Since the runtime target was already met,
the accuracy margin was kept. `nCorrectors 2` in the fluid `fvSolution` was
also tested and has *exactly* zero effect on the corrector count — the inner
PIMPLE work is not the bottleneck.

**Figure 6 and the reported timings are regenerated** from a full cardiac
cycle with the shipped settings:

- 20,000 time steps in 5633 s of solver time on 8 cores
- mean 1.28 FSI correctors per time step; 77% of steps converge in a single
  corrector, 19% in two
- 11 correctors in the first time step (unchanged), max 11 over the cycle
- `nOuterCorr 30` never reached; every step ended in convergence state 1

### Caveats

- The previous README figure of 9209 s was measured under **v2512**; this run
  is **v2412**, so the wall-clock comparison mixes the version change with the
  tolerance and `hsModel` changes and is not a clean attribution. The README
  now states v2412.
- Solution agreement between the shipped tolerances and the defaults was
  verified over 100 steps (3.5e-4 relative displacement, 1.7e-3 force) and by
  the regression test, **not** by a full-cycle comparison at default
  tolerances.
- `robinCouplingStudy/RESULTS.md` also documented the old default, but that
  file is removed on this branch by 46aa02d9, so that edit was dropped.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by conor-moore-123** [comment]
### Independent review of 507914b2 (`enhancedVolPointInterpolation`: store the old-time point field)

I reviewed this commit against a separate investigation of the same defect, reached from a different direction: three patient-specific arterial FSI cases where the fluid mesh was realising only 1–9% of the solid interface motion. Same root cause, and **this fix is correct**. A few notes that may be useful.

**What I checked**

- **Overload resolution does route existing callers to the safe path.** The four call sites — `linGeomTotalDispSolid.C:326`/`:426` and `nonLinGeomTotalLagTotalDispSolid.C:507`/`:617` — pass `(volVectorField, volTensorField, pointVectorField)`. The old overload is viable only via derived-to-base conversion on all three arguments; the new one is an exact match on all three, so it wins unambiguously with no source changes. `Type` deduces consistently from arguments 1 and 3; argument 2's `outerProduct<vector,Type>::type` is a non-deduced context, which is harmless here.
- **The save is armed from the first time step.** `storeOldTimes()` only acts when `field0Ptr_` is non-null, and `solidModel.C:1230` forces `pointD_.oldTime()` in the constructor. (Worth stating explicitly — a `ref()`-based fix would be a no-op on step 1 otherwise.)
- **The multi-material path was already safe**, as the commit message says: `solidSubMeshes::mapSubMeshPointFields` writes via `baseMeshField.ref()`, and the sub-mesh interpolation passes a `pointVectorField`, so it picks up the new overload too.

**Independent reproduction**, on a different case family from the one in the commit message — Robin–Neumann, `fixedRelaxation` with `relaxationFactor 1`, `linearGeometryTotalDisplacement`, 32 MPI ranks, 439k fluid cells. Per-step ratio of fluid wall motion to solid `pointD` motion, area-weighted over 73,344 conformal interface points:

| step | unmasked case | with a `fixedDisplacement`/`fixedValue` pair present |
|---|---|---|
| 1 | 1.0000 | 1.0000 |
| 2 | 0.5141 | 1.0000 |
| 3 | 0.3164 | 1.0000 |
| 4 | 0.2166 | 1.0000 |
| 5 | 0.1592 | 1.0000 |

The same ~1/n decay. Bit-identical results with `diffusivity uniform`, `interpolatorUpdateFrequency 1` and `predictor no`, which places the loss upstream of the coupling, the mesh motion and the AMI — consistent with the diagnosis here. Direct measurement with a coded function object gave `lostPart = pointD.oldTime() − previous pointD` **exactly 0** in 3dTube as shipped and non-zero and growing (48% → 77% of the true increment over two steps) in an unmasked case.

**Two suggestions, neither a defect in this change**

1. **The base-class overload is still `public`** (`enhancedVolPointInterpolation.H:187`). It can't be removed — the new wrapper calls it — but it stays reachable by any future caller whose third argument is statically a `DimensionedField&`, and that recurrence would be silent: no compiler diagnostic (binding derived to base is legal), no runtime error (values are correct, only the archived copy is stale), and the symptom surfaces as a wrong subtraction several frames away. That is the failure mode that survived here since #126. Making it `protected`/`private` would cost nothing internally — it has exactly one in-repo caller — or failing that a `\warning` on the signature pointing at the new overload.
2. **A regression test is worth adding**, because the shipped suite structurally cannot catch this. All nine FSI tutorials carry the masking combination (a `fixedDisplacement` `D` patch backed by a `fixedValue` `pointD` patch), `3dTube/regressionTest.sh` validates the *solid's* point displacement — which is correct under the defect — and its reference values are self-generated, so it detects changes from a baseline rather than a defect already present when the baseline was recorded. The reproducer in the commit message is the right shape: 3dTube with the inlet/outlet `pointD` patches changed `fixedValue` → `calculated`, asserting the **per-step** ratio at every step over ~10 steps. Per-step rather than cumulative, because step 1 reads 1.000 even with the defect. `|1 − ratio| < 1e-3` is comfortable: measured scatter with the fix is ~3e-6, and the defect puts step 2 at 0.489.

**Not tested by me:** I did not rebuild or run the patched library, so the clean build and the "tutorial unchanged to the printed digits" claim are unverified on my side. The foam-extend interpolator path is unchecked. The non-linear total-Lagrangian model and the segregated path go through the same overload resolution, so I'd expect them covered by construction, but neither was run.

File: src/solids4FoamModels/fluidSolidInterfaces/fluidSolidInterface/fluidSolidInterface.C [hunk 1]
```diff
@@ -1618,10 +1770,566 @@ Foam::scalar Foam::fluidSolidInterface::updateResidual()
...
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+
+namespace Foam
```
Comment (iBatistic) [inline src/solids4FoamModels/fluidSolidInterfaces/fluidSolidInterface/fluidSolidInterface.C]: Can we put everything in namespace Foam. We usually do not put only one function in namespace like this.

File: src/solids4FoamModels/fluidSolidInterfaces/fluidSolidInterface/fluidSolidInterface.H [hunk 2]
```diff
@@ -154,6 +154,102 @@ class fluidSolidInterface
...
+        //- Largest pressure norm in the current time step per interface
+        List<scalar> maxRobinPressureNorm_;
+
+
+        //- Current maximum normalized Robin pressure residual
+        scalar robinPressureResidual_;
+
+        //- Current maximum normalized Robin leakage-flux residual
+        scalar robinFluxResidual_;
+
```
Comment (iBatistic) [inline src/solids4FoamModels/fluidSolidInterfaces/fluidSolidInterface/fluidSolidInterface.H]: Formating: two lines instead of one line (at different locations).

File: tutorials/fluidSolidInteraction/beamInCrossFlow/verification/scripts/beam_in_cross_flow_verification.py [hunk 3]
```diff
@@ -355,26 +447,116 @@ def write_results(name: str, rows: list[dict], references: dict, order: float |
...
+
+
+def run_coupling_study(args: argparse.Namespace, references: dict) -> bool:
+    mesh_spec = references[args.case].get("mesh", references["mesh"])
+    delta_t = mesh_spec["deltaTs"][0]
+    end_time = mesh_spec["endTime"]
+    cores = study_cores(args.cores, args.case, 1)
+    results: dict[str, dict[str, float]] = {}
+
+    for coupling in ("iqnils", "robin"):
```
Comment (iBatistic) [inline tutorials/fluidSolidInteraction/beamInCrossFlow/verification/scripts/beam_in_cross_flow_verification.py]: Extracts each run’s last available result without checking that displacement, force, and residual histories reach the configured end time.

Fix: require successful completion and matching displacement/force/residual timestamps at the requested end time before comparing values.

File: src/solids4FoamModels/fluidModels/fvPatchFields/elasticWallPressure/elasticWallPressureFvPatchScalarFieldRobinCoeff.C [hunk 4]
```diff
@@ -0,0 +1,1190 @@
...
+{
+    const labelListList& faceFaces = patch().patch().faceFaces();
+
+    for (label sweepI = 0; sweepI < nSweeps; sweepI++)
+    {
+        const scalarField fld0(fld);
+
+        forAll(fld, faceI)
+        {
+            const labelList& nbrs = faceFaces[faceI];
```
Comment (iBatistic) [inline src/solids4FoamModels/fluidModels/fvPatchFields/elasticWallPressure/elasticWallPressureFvPatchScalarFieldRobinCoeff.C]: patch-local faceFaces() without exchanging neighbours across processor boundaries.

two adjacent faces with values [1, 3] smooth to [2, 2] together, but remain [1, 3] when partitioned into separate patches. Consequently, secantFit median with secantSpatial yes produces decomposition-dependent coefficient updates. The default minimax configuration is unaffected.

Fix: smooth on the global interface or exchange neighbouring values across processor boundaries.

File: tutorials/fluidSolidInteraction/beamInCrossFlow/constant/fsiProperties.robin [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    // time-step, then the IQNILS procedure is used
+    relaxationFactor 1.0;
+
+    // Tolerance for the FSI loop within each time-step
+    outerCorrTolerance 1e-6;
+
+    // Additional convergence tolerances used automatically by Robin pressure
+    // interfaces: interface-pressure change and kinematic (interface
+    // leakage-flux change) residuals; see README.md for the iteration-error
+    // estimate (robinConvergence iterationError) used to test them
```
Comment (iBatistic) [inline tutorials/fluidSolidInteraction/beamInCrossFlow/constant/fsiProperties.robin]: Describes leakage-flux change and implies iterationError is active; the default now tests leakage using residual

**Comment by philipcardiff** [comment]
@conor-moore-123 thanks, this was a very useful independent check, especially the second reproduction on a different case family and the `lostPart` measurement. I've acted on both suggestions in 486a0214.

**1. The base-class overload is now private.** The `DimensionedField` overload of `interpolate(vf, gradVf, pf)` is moved to the private section, with a comment explaining why. Its only caller is the new `GeometricField` wrapper, and `Allwmake` builds cleanly on v2412. A future call that binds a `GeometricField` to it now fails to compile instead of silently leaving `oldTime()` stale.

**2. Regression test.** `3dTube/regressionTest.sh` now runs a second copy of the case for 10 steps with the inlet/outlet `pointD` patches changed from `fixedValue` to `calculated` (`D` still `fixedDisplacement`), as you suggested. A coded function object (`system/interfaceMotionRatio`) prints the per-step ratio of fluid `wall` point motion to the solid `inner-wall` `pointD` increment. It stores its own previous values rather than reading `pointD.oldTime()` or `oldPoints()`, so the test doesn't depend on the code it's testing. Steps 2–10 must satisfy `|1 − r| < 1e-3`, and all 9 steps must report.

Results on OpenFOAM-v2412:

| run | existing disp/force checks | interface motion ratio, steps 2–10 |
|---|---|---|
| with the fix | pass | 1.000000 to 0.999999 → pass |
| fix reverted (wrapper passes `pf` as `DimensionedField&`) | **still pass** | 0.532, 0.340, 0.230, … 0.050 → **fail** |

The second row confirms your point that the existing checks couldn't see this.

**Limitations:** the check is skipped on foam-extend, which has no coded function objects. It hasn't been run on OpenFOAM.org, though it uses only standard API. It assumes a serial run and a conformal interface, both of which hold for 3dTube.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: src/solids4FoamModels/fluidSolidInterfaces/fluidSolidInterface/fluidSolidInterface.C (hunk 1 above)
Comment (philipcardiff) [reply src/solids4FoamModels/fluidSolidInterfaces/fluidSolidInterface/fluidSolidInterface.C]: Done in c678a1aa: `robinResidualTest` now sits in the anonymous namespace inside the top `namespace Foam` block, next to the other file-local helpers (`resolvedRegion`, `warnOnConflictingRegion`). No functional change.

File: src/solids4FoamModels/fluidSolidInterfaces/fluidSolidInterface/fluidSolidInterface.H (hunk 2 above)
Comment (philipcardiff) [reply src/solids4FoamModels/fluidSolidInterfaces/fluidSolidInterface/fluidSolidInterface.H]: Done in c678a1aa: removed the four double blank lines this PR added in the header.

File: tutorials/fluidSolidInteraction/beamInCrossFlow/verification/scripts/beam_in_cross_flow_verification.py (hunk 3 above)
Comment (philipcardiff) [reply tutorials/fluidSolidInteraction/beamInCrossFlow/verification/scripts/beam_in_cross_flow_verification.py]: Done in 3a76d0ef. `run_case` now fails unless `solids4Foam` printed `End`. `extract` and `robin_residual_summary` also fail unless the displacement, force and Robin residual histories reach the configured end time (`math.isclose`, rel. tol. 1e-6), so values from a truncated run are no longer compared. On a finished case the check accepts t = 4 and rejects a requested t = 5.

File: src/solids4FoamModels/fluidModels/fvPatchFields/elasticWallPressure/elasticWallPressureFvPatchScalarFieldRobinCoeff.C (hunk 4 above)
Comment (philipcardiff) [reply src/solids4FoamModels/fluidModels/fvPatchFields/elasticWallPressure/elasticWallPressureFvPatchScalarFieldRobinCoeff.C]: Good catch, fixed in 6fcdf3ea. `smoothPatchField` now maps the field to the global interface patch (`globalPolyPatch::patchFaceToGlobal`), smooths there, and maps back. The global patch is the same on every processor and connected across processor boundaries (duplicate points are merged), so the smoothing no longer depends on the decomposition. Serial results are bitwise unchanged.

Test: beamInCrossFlow Robin with `secantFit median; secantSpatial yes;`, 4 cores (interface split 128/64 faces) vs serial. Largest relative difference in the mean secant scale:

| | first 50 updates | first 200 updates |
|---|---|---|
| before | 3.4e-2 | 3.4 |
| after | 2.4e-4 | 9.6e-3 |

The remainder is parallel round-off, the same order as with the default `minimax` fit, which doesn't use the smoothing.

File: tutorials/fluidSolidInteraction/beamInCrossFlow/constant/fsiProperties.robin (hunk 5 above)
Comment (philipcardiff) [reply tutorials/fluidSolidInteraction/beamInCrossFlow/constant/fsiProperties.robin]: Done in 3a76d0ef: the comment now describes the default `robinConvergence residual` test of the pressure change and the leakage (when the fluid mesh follows the solid), and presents `iterationError` as optional.
